### DISCUSSION TITLE
Creating packages on conda-forge

### DISCUSSION THREAD

**Discussion by maresb** [pyscaffold/pyscaffold#422] [Best Practices & Workflows]
I just [added a conda-forge package for dsproject](https://github.com/pyscaffold/pyscaffoldext-dsproject/issues/7). conda-forge is the leading package management system for Python-centric data science. With the recent release of the C++ remake of `conda` named [`mamba`](https://github.com/mamba-org/mamba), the performance problems have been solved, and the `conda` ecosystem has a bright future.

conda-forge is the leading open-source conda repository. Publishing to conda-forge was quite a bit more work than I initially expected, but in the end once one understands what's going on, it's not so terribly complicated. It's however absolutely essential that one is proficient with Git, in particular with forks, branches, and Github pull requests.

This is a quick brain-dump of what I learned during my first conda-forge packaging adventure. This attempts to be neither complete nor accurate, just a summary of the most important concepts. I assume Linux, but probably OSX works the same. Corrections and additions are most welcome.

Let's start by looking [here at `meta.yaml`](https://github.com/conda-forge/pyscaffoldext-dsproject-feedstock/blob/master/recipe/meta.yaml). Everything is controlled by this single file. All the other files in the repo are rendered automatically.

There is a tool called [grayskull](https://conda-forge.org/blog/posts/2020-03-05-grayskull/) which attempts to generate a viable `meta.yaml`. Run
```bash
conda install -c conda-forge grayskull
grayskull pypi pyscaffoldext-dsproject
```
That will create a `pyscaffoldext-dsproject/meta.yaml`.

By default this creates a package for each architecture, which is overkill for a pure Python project. Thus we should add `noarch: python` under `build:`. Then we are creating an architecture-independent package.

With a `noarch` package, one must specify a minimum Python version in both host and run requirements, so in both places we change `- python` to `- python >=3.6`, taking care that there is a space between `python` and `>=3.6` but no space between `>=` and `3.6`.

Also with a `noarch` package, one is not allowed to use platform-specific selectors (metacommands) like `# [win]` or even `# [py<38]`. Thus we must remove them.

There is a conda support package called [`importlib-metadata`](https://pypi.org/project/importlib-metadata/) which is a backport of some Python 3.8 functionality. Since we are using `noarch` we have to include the conda package regardless of Python version. It's not completely optimal, but it's also not harmful.

Grayskull misses our host requirement for `setuptools-scm`. If I build my package without `setuptools-scm`, then the package version cannot be read from version control, and will be set as `0.0.0` in the wheel. (The conda and pip version numbers should be identical, but this is not enforced.) Everything will build successfully. Other conda packages will also be able to install and import my package as a dependency. However, `pip` will be left in an inconsistent state if any version requirement is specified for my package. (Pip requirements need `mypackage>=1.2.3` but the installed version is `0.0.0`.)

Note that there are two conda packages for `setuptools-scm`. The primary one is `setuptools-scm` but there is a wrapper package called `setuptools_scm` which depends on `setuptools-scm`. To avoid having both installed, it's better to install `setuptools-scm`.

Finally, we should add some test to make sure that the program executes after installing. In this case, we simply add `putup --help` to the test commands. This command should be runnable from Linux, OSX, and Windows, so it can't be anything fancy.

Finally, we have to get our `meta.yaml` recipe into conda-forge. We do this by forking [staged-recipes](https://github.com/conda-forge/staged-recipes), making a new local branch for our package, placing the `meta.yaml` in a subdirectory of the `recipes` directory, and making a pull request. Detailed documentation is found under [contributing packages](https://conda-forge.org/docs/maintainer/adding_pkgs.html) and [knowledge base](https://conda-forge.org/docs/maintainer/knowledge_base.html). The part of the instructions for creating the `meta.yaml` are superceeded by the instructions here, where we instead used Grayskull.

Once a pull request is made, the recipe will be linted and then test-built on linux/osx/win. If everything passes (or if you get stuck even after a good-faith effort to find answers in the docs), then you can get human review for a pure-Python package by writing a message to @conda-forge/help-python in the PR, or by contacting the team on [Gitter](https://gitter.im/conda-forge/conda-forge.github.io).

Once your recipe is approved it will be converted to a feedstock with its own repository. You should only edit the feedstock by PRs (otherwise conda-forge will generate extraneous builds for your package, and such builds are permanent).

You can check the build status [on anaconda.org](https://anaconda.org/conda-forge/pyscaffoldext-dsproject/files). Once a build is generated, it will appear here almost immediately. However, it takes roughly an hour before it's available from the `conda` command itself. This is due to the CDN which mirrors the files. You can check the availablility of files on the CDN with `conda search`:

```
$ conda search pyscaffoldext-dsproject
Loading channels: done
# Name                       Version           Build  Channel             
pyscaffoldext-dsproject             0.6  py36h5fab9bb_0  conda-forge         
pyscaffoldext-dsproject             0.6  py37h89c1867_0  conda-forge         
pyscaffoldext-dsproject             0.6  py38h578d9bd_0  conda-forge         
pyscaffoldext-dsproject             0.6    pyhd8ed1ab_0  conda-forge         
pyscaffoldext-dsproject             0.6    pyhd8ed1ab_1  conda-forge         
```

Here you can see that for my first build I forgot `noarch`, so separate builds for Python 3.6, 3.7, and 3.8 were generated. Additional builds appear in Anaconda.org, and I suspect this is due to me accidentally committing directly to the feedstock instead of to my fork. Try not to make mistakes, but when you do inevitably make a mess, the team on Gitter is in my experience friendly and understanding, and will help you sort it out.

The `_#` suffix on the build indicates the build number, which is for tracking when the version number remains the same, but the conda package is modified due to a bugfix. It's like an extra semver digit: it should be bumped as appropriate, and reset to 0 when the version number changes.

Once your recipe graduates to a feedstock, assuming it is `noarch` it will only be tested against Linux. That means that you can use non-Windows stuff like `grep` in your test commands. (I asked on Gitter whether this is bad practice, and nobody responded, so I assume that nobody was horrified.)

There is also a question of whether unit tests should be included in the conda-forge tests. It seems to be that running unit tests is discouraged as something that should be done upstream, but it's not forbidden.

Making a feedstock requires quite a bit of patience and persistence due to various subtleties and idiosyncrasies, for example, one should [almost never depend on `matplotlib` but instead on `matplotlib-base`](https://conda-forge.org/docs/maintainer/knowledge_base.html#matplotlib).

The building process is quite slow, but the experimental new [boa](https://github.com/mamba-org/boa) package builder promises to accelerate the process in the future. But since I never explicitly ran `conda-build`, I never explored the possibility myself.

That's all for now, happy packaging!

**Comment by abravalheri** [comment]
Thank you very much @maresb for the awesome tips... If you are not against it, I think we probably will include a link for this discussion somewhere in the website.

**Reply by maresb** [reply]
I give permission for anyone to do whatever they want with the above writeup. So feel free to link, copy, rewrite, etc. as wanted.

**Comment by mfhepp** [comment]
I am cross-referencing #633 here because it contains useful information about adding `conda-forge` deployment support to PyScaffold.
